### PR TITLE
Dact 332 transaction interceptor

### PR DESCRIPTION
**Pull request by KALatchman**
In order to check the transaction is valid, we will need to build a Transaction interceptor
 
TransactionsInterceptor:

The officer API will need to pass on any errors from the Transactions API.

The officer API will need to check - in a transaction interceptor - whether it’s appropriate to make further changes to its own submission data by checking the Transaction. If the Transaction is in any state other than OPEN then no new data should be written to the collection. This doesn’t preclude any internal updates we might wish to make, based on any internal events, this refers to incoming calls only.

TransactionID removed as transactions is now a request attribute, in order for the interceptor to check if a transaction is valid.

 CURRENT ERROR MESSGE IS A PLACEHOLDER

https://companieshouse.atlassian.net/jira/software/c/projects/DACT/boards/477?modal=detail&selectedIssue=DACT-332

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/officerfiling/api/interceptor/TransactionInterceptorTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Test
+    void testPreHandleIsSuccessful() {
+        transaction.setStatus(TransactionStatus.OPEN);
+        when(transactionService.getTransaction(TRANS_ID, PASSTHROUGH_HEADER)).thenReturn(transaction);
+        when(mockHttpServletRequest.getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE)).thenReturn(pathParams);
+        when(mockHttpServletRequest.getHeader("ERIC-Access-Token")).thenReturn(PASSTHROUGH_HEADER);
+
+        assertTrue(transactionInterceptor.preHandle(mockHttpServletRequest, mockHttpServletResponse, mockHandler));
+        verify(mockHttpServletRequest, times(1)).setAttribute("transaction", transaction);
```
Comment (jditchburn) [inline src/test/java/uk/gov/companieshouse/officerfiling/api/interceptor/TransactionInterceptorTest.java:65]: You only have to specify the number of times when it is not 1. If you remove that argument from the method, the default is 1 anyway.

File: src/main/java/uk/gov/companieshouse/officerfiling/api/utils/Constants.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.companieshouse.officerfiling.api.utils;
+
+public class Constants {
```
[the remark this answers is not in the document]
Comment (KALatchman) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/utils/Constants.java:3]: I spoke to Richard about including a class to use frequent strings that should be constants. So for example: 
public static final String SUBMISSION_URI_PATTERN = "/transactions/%s/officers";

[the remark this answers is not in the document]
Comment (rwalrondch) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/utils/Constants.java:3]: Yes anywhere where we are using a string in multiple places, made sense to put it in a constants class, so if it changes in the future, we just have to change it in one place.

**Comment by PhilipGlover** [comment]
We've only done a 'spike' for this so far, but we used the Interceptors in this library - https://github.com/companieshouse/api-security-java/tree/master/src/main/java/uk/gov/companieshouse/api/interceptor

**Review by PhilipGlover — APPROVED** [review]
- have run successfully, can file Officer details as normal
- if I manually close the transaction in the db I get
```
    "errors": [
        {
            "error": "Error Updating Transaction 108423-460616-748084",
            "error_values": {
                "cause": "403 Forbidden\nCannot Update/Delete a transaction with CLOSED/DELETED status"
            },
            "location": "/transactions/108423-460616-748084/officers",
            "location_type": "resource",
            "type": "ch:service"
        }
    ]
```
- all test pass, I haven't looked carefully at the tests, but adding Mock Beans for the Interceptors is what we've done too.  In fact we've started to do a Base Test class to do these common setup things.

File: src/main/java/uk/gov/companieshouse/officerfiling/api/config/InterceptorConfig.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+import uk.gov.companieshouse.api.interceptor.OpenTransactionInterceptor;
+import uk.gov.companieshouse.api.interceptor.TransactionInterceptor;
+
+@Configuration
+@ComponentScan("uk.gov.companieshouse.api")
+public class InterceptorConfig implements WebMvcConfigurer {
+
+    private static final String TRANSACTIONS = "/transactions/**";
+    static final String[] TRANSACTIONS_LIST = {TRANSACTIONS, "/private/**"};
```
Comment (jditchburn) [inline src/main/java/uk/gov/companieshouse/officerfiling/api/config/InterceptorConfig.java:17]: Can this not be private?

**Review by jditchburn — APPROVED** [review]

**Pull request merged**